### PR TITLE
[ONNX] Improve diagnostic message formatting

### PR DESCRIPTION
**Pull request by BowenBao**
* Reflect required arguments in method signature for each diagnostic rule. Previous design accepts arbitrary sized tuple which is hard to use and prone to error.
     image
* Removed `DiagnosticTool` to keep things compact.
* Removed specifying supported rule set for tool(context) and checking if rule of reported diagnostic falls inside the set, to keep things compact.
* Initial overview markdown file.
* Change `full_description` definition. Now `text` field should not be empty. And its markdown should be stored in `markdown` field.
* Change `message_default_template` to allow only named fields (excluding numeric fields). `field_name` provides clarity on what argument is expected.
* Added `diagnose` api to `torch.onnx._internal.diagnostics`. 

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #88261
* #87258
* #88145
* __->__ #87830

### CODE REVIEW

File: torch/onnx/_internal/diagnostics/infra/_infra.py [hunk 1]
```diff
@@ -32,36 +32,48 @@ class Tag(enum.Enum):
...
-            kwargs["full_description"]["markdown"]
-            if "full_description" in kwargs
-            else None
+        short_description = kwargs.get("short_description", {}).get("text", None)
+        full_description = kwargs.get("full_description", {}).get("text", None)
+        full_description_markdown = kwargs.get("full_description", {}).get(
+            "markdown", None
         )
-        help_uri = kwargs["help_uri"] if "help_uri" in kwargs else None
+        help_uri = kwargs.get("help_uri", None)
```
Comment (justinchuby) [inline torch/onnx/_internal/diagnostics/infra/_infra.py:68]: I think get returns a None if not found by default? `kwargs.get("help_uri")`

File: torch/onnx/_internal/diagnostics/infra/_infra.py [hunk 2]
```diff
@@ -147,21 +170,42 @@ def add_frame(
...
+    graphs: List[Graph] = dataclasses.field(default_factory=list)
     additional_message: Optional[str] = None
     tags: List[Tag] = dataclasses.field(default_factory=list)
 
     def sarif(self) -> sarif.Result:
         """Returns the SARIF Result representation of this diagnostic."""
-        if self.message_args is None:
-            self.message_args = tuple()
-        message = self.rule.message_default_template.format(*self.message_args)
+        message = self.message
```
Comment (justinchuby) [inline torch/onnx/_internal/diagnostics/infra/_infra.py:206]: minor:

```
message = self.message or self.rule.message_default_template
```

simpler but not necessary

File: torch/onnx/errors.py [hunk 3]
```diff
@@ -46,49 +46,27 @@ class UnsupportedOperatorError(OnnxExporterError):
...
-                diagnostics.levels.ERROR,
-                message_args=(
-                    name,
-                    version,
-                    supported_version,
-                ),
+            diagnostic_rule: diagnostics.infra.Rule = (
+                diagnostics.rules.operator_supported_in_newer_opset_version
             )
+            msg = diagnostic_rule.format_message(name, version, supported_version)
```
Comment (justinchuby) [inline torch/onnx/errors.py:53]: Love this

**Review by abock — APPROVED** [review]

**Pull request closed without merging**